### ISSUE TITLE
Duplicate publisher jobs after unregister and new discovery

### ISSUE THREAD

**Issue by martin-weber** [Azure/Industrial-IoT#685]
## Description of the Bug

When unregistering an asset in the Engineering Tool or via swagger (DELETE /registry/v2/applications/\{applicationId\}) the Jobs for the endpoint are not deleted.
This results in following consequences:
- Nodes/Tags are still processed on the edge publisher module.
- When running a new discovery and registering the same endpoint again, it results in a second dataSetWriter in the job in Cosmos DB with the same dataSetWriterId, when one of the alternative `connection.endpoint.alternativeUrls` change between the discoveries.

## Steps to reproduce

I used the free Pro-SYS OPC-UA Simulation Server to reproduce this issue. These are the steps:

* Start OPC-UA Simulation server. And disable some of the binded addresses in the Tab Endpoints -> Bind Addresses -> Custom
*  In the Engineering Tool run a discovery and turn on one of the asset/endpoint
* Turn on publishing of an endpoint item
* Wait for the job to run
* check cosmos DB entry for the Job entry. It contains now one dataSetWriter for the given endpoint. In my example with the following dataSetWriterId:
  ```json
  "dataSetWriterId": "uat442d2f329e1ba4e37320a299ae18564d8885982d",
  ```
  and the following connection:
  ```json
                                "connection": {
                                    "endpoint": {
                                        "url": "opc.tcp://192.168.1.146:53530/OPCUA/marweber",
                                        "alternativeUrls": [
                                            "opc.tcp://172.23.112.1:53530/OPCUA/marweber",
                                            "opc.tcp://host.docker.internal:53530/OPCUA/marweber"
                                        ],
                                        "securityMode": "None",
                                        "securityPolicy": "http://opcfoundation.org/UA/SecurityPolicy#None",
                                        "certificate": "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
                                    }
                                },
  ```
* Turn off discover
* Unregister the asset 
* **Questions:**
  * The CosmosDB entry for the job still remains. **Is this expected?**
  * If running the module form visual studio, the debug output still seams to read from the server, as it prints messages like this:  `NotificationReceived: ClientHandle=175, Value=-86.66667` **Is this expected, or should it stop?**

* In the OPC-UA Simulation Server in the Tab Endpoints -> Bind Addresses -> Custom activate one of the previously deactivated addresses. Then apply and restart the OPC-UA server
* In the engineering portal start a new discovery.
* Add the same endpoint again.
* Check the CosmosDB entry. A new data set writer with the same dataSetWriterId is created
  ```json
   "dataSetWriterId": "uat442d2f329e1ba4e37320a299ae18564d8885982d" 
   ```
   but with the new connection, even though the connection.endpoitn.url is still the same
  ```json
                                "connection": {
                                    "endpoint": {
                                        "url": "opc.tcp://192.168.1.146:53530/OPCUA/marweber",
                                        "alternativeUrls": [
                                            "opc.tcp://172.23.112.1:53530/OPCUA/marweber",
                                            "opc.tcp://host.docker.internal:53530/OPCUA/marweber",
                                            "opc.tcp://192.168.64.1:53530/OPCUA/marweber",
                                            "opc.tcp://100.X.X.X:53530/OPCUA/marweber"
                                        ],
                                        "securityMode": "None",
                                        "securityPolicy": "http://opcfoundation.org/UA/SecurityPolicy#None",
                                        "certificate": "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
                                    }
                                },
  ```
* When turning on the asset again and navigation to the corresponding endpoint control in the engineering portal, the item is already / still turned on and shows updating data.

**Questions**:
  * I would expect that the jobs of end endpoint are deleted when the endpoint is unregistered. This would also solve the issues that the rediscovery adds a new entry. Is this correct?
  * The Visual studio output  prints still the `NotificationReceived: ClientHandle=175, Value=-86.66667` messages even when no endpoint is turned on since the discovery of the endpoint. Shouldn't this stop when the asset is unregistered?
 

## Final Job entry in CosmosDB

This is the final entry in CosmosDB after the 
```json
{
    "id": "uat442d2f329e1ba4e37320a299ae18564d8885982d",
    "_etag": "\"9302a860-0000-0200-0000-5f460f620000\"",
    "ClassType": "Job",
    "JobId": "uat442d2f329e1ba4e37320a299ae18564d8885982d",
    "Name": "uasc6a5294fa2350663f3c53dbca9b8c6e880c36f4e",
    "Type": "DataSetWriterV2",
    "JobConfiguration": {
        "id": null,
        "_etag": null,
        "ClassType": "JobConfig",
        "JobId": "uat442d2f329e1ba4e37320a299ae18564d8885982d",
        "Job": {
            "writerGroup": {
                "writerGroupId": "opc.tcp://host.docker.internal:53530/OPCUA/marweber",
                "messageType": "Json",
                "dataSetWriters": [
                    {
                        "dataSetWriterId": "uat442d2f329e1ba4e37320a299ae18564d8885982d",
                        "dataSet": {
                            "dataSetSource": {
                                "publishedVariables": {
                                    "publishedData": [
                                        {
                                            "publishedVariableNodeId": "http://www.prosysopc.com/OPCUA/SampleAddressSpace#s=MyLevel",
                                            "samplingInterval": "00:00:01",
                                            "queueSize": 1
                                        },
                                        {
                                            "publishedVariableNodeId": "http://www.prosysopc.com/OPCUA/SimulationNodes/#i=1004",
                                            "samplingInterval": "00:00:01",
                                            "queueSize": 1
                                        }
                                    ]
                                },
                                "connection": {
                                    "endpoint": {
                                        "url": "opc.tcp://192.168.1.146:53530/OPCUA/marweber",
                                        "alternativeUrls": [
                                            "opc.tcp://172.23.112.1:53530/OPCUA/marweber",
                                            "opc.tcp://host.docker.internal:53530/OPCUA/marweber"
                                        ],
                                        "securityMode": "None",
                                        "securityPolicy": "http://opcfoundation.org/UA/SecurityPolicy#None",
                                        "certificate": "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
                                    }
                                },
                                "subscriptionSettings": {
                                    "publishingInterval": "00:00:01",
                                    "resolveDisplayName": true
                                }
                            },
                            "dataSetMetaData": {
                                "name": "uat442d2f329e1ba4e37320a299ae18564d8885982d",
                                "dataSetClassId": "1a83449e-2aa0-43df-93f7-bab946ccd8ab",
                                "configurationVersion": {
                                    "majorVersion": 1,
                                    "minorVersion": 27
                                }
                            },
                            "extensionFields": {
                                "EndpointId": "uat442d2f329e1ba4e37320a299ae18564d8885982d",
                                "PublisherId": "uat442d2f329e1ba4e37320a299ae18564d8885982d",
                                "DataSetWriterId": "uat442d2f329e1ba4e37320a299ae18564d8885982d"
                            }
                        },
                        "dataSetFieldContentMask": "StatusCode, SourceTimestamp, ServerTimestamp, NodeId, DisplayName, EndpointUrl, ApplicationUri, ExtensionFields",
                        "messageSettings": {
                            "dataSetMessageContentMask": "Timestamp, MetaDataVersion, Status, DataSetWriterId, MajorVersion, MinorVersion, SequenceNumber"
                        }
                    },
                    {
                        "dataSetWriterId": "uat442d2f329e1ba4e37320a299ae18564d8885982d",
                        "dataSet": {
                            "dataSetSource": {
                                "publishedVariables": {
                                    "publishedData": [
                                        {
                                            "publishedVariableNodeId": "http://www.prosysopc.com/OPCUA/SimulationNodes/#i=1004",
                                            "samplingInterval": "00:00:01",
                                            "queueSize": 1
                                        }
                                    ]
                                },
                                "connection": {
                                    "endpoint": {
                                        "url": "opc.tcp://192.168.1.146:53530/OPCUA/marweber",
                                        "alternativeUrls": [
                                            "opc.tcp://172.23.112.1:53530/OPCUA/marweber",
                                            "opc.tcp://host.docker.internal:53530/OPCUA/marweber",
                                            "opc.tcp://192.168.64.1:53530/OPCUA/marweber",
                                            "opc.tcp://100.X.X.X:53530/OPCUA/marweber"
                                        ],
                                        "securityMode": "None",
                                        "securityPolicy": "http://opcfoundation.org/UA/SecurityPolicy#None",
                                        "certificate": "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
                                    }
                                },
                                "subscriptionSettings": {
                                    "publishingInterval": "00:00:01",
                                    "resolveDisplayName": true
                                }
                            },
                            "dataSetMetaData": {
                                "name": "uat442d2f329e1ba4e37320a299ae18564d8885982d",
                                "dataSetClassId": "3ad2a4b0-8895-4aa6-ad41-4adff4a6f9d8",
                                "configurationVersion": {
                                    "majorVersion": 1,
                                    "minorVersion": 1
                                }
                            },
                            "extensionFields": {
                                "EndpointId": "uat442d2f329e1ba4e37320a299ae18564d8885982d",
                                "PublisherId": "uat442d2f329e1ba4e37320a299ae18564d8885982d",
                                "DataSetWriterId": "uat442d2f329e1ba4e37320a299ae18564d8885982d"
                            }
                        },
                        "dataSetFieldContentMask": "StatusCode, SourceTimestamp, ServerTimestamp, NodeId, DisplayName, EndpointUrl, ApplicationUri, ExtensionFields",
                        "messageSettings": {
                            "dataSetMessageContentMask": "Timestamp, MetaDataVersion, Status, DataSetWriterId, MajorVersion, MinorVersion, SequenceNumber"
                        }
                    }
                ],
                "messageSettings": {
                    "networkMessageContentMask": "PublisherId, WriterGroupId, NetworkMessageNumber, SequenceNumber, PayloadHeader, Timestamp, DataSetClassId, NetworkMessageHeader, DataSetMessageHeader"
                }
            },
            "connectionString": "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX",
            "messagingMode": "Samples",
            "engine": {
                "batchSize": 50,
                "batchTriggerInterval": "00:00:10",
                "diagnosticsInterval": "00:01:00",
                "maxMessageSize": 0
            }
        }
    },
    "Demands": [
        {
            "id": null,
            "_etag": null,
            "ClassType": "Demand",
            "JobId": "uat442d2f329e1ba4e37320a299ae18564d8885982d",
            "Key": "Type",
            "Operator": null,
            "Value": "publisher"
        },
        {
            "id": null,
            "_etag": null,
            "ClassType": "Demand",
            "JobId": "uat442d2f329e1ba4e37320a299ae18564d8885982d",
            "Key": "SiteId",
            "Operator": null,
            "Value": "marweber-dev"
        }
    ],
    "DesiredActiveAgents": 1,
    "DesiredPassiveAgents": 0,
    "Status": "Active",
    "ProcessingStatus": {
        "marweber-dev_publisher_1": {
            "id": null,
            "_etag": null,
            "ClassType": "ProcessingStatus",
            "JobId": "uat442d2f329e1ba4e37320a299ae18564d8885982d",
            "LastKnownHeartbeat": "2020-08-25T15:32:34.5510992Z",
            "LastKnownState": null,
            "ProcessMode": "Active"
        },
        "marweber-dev_publisher_2": {
            "id": null,
            "_etag": null,
            "ClassType": "ProcessingStatus",
            "JobId": "uat442d2f329e1ba4e37320a299ae18564d8885982d",
            "LastKnownHeartbeat": "2020-08-26T07:29:38.0799652Z",
            "LastKnownState": null,
            "ProcessMode": "Active"
        }
    },
    "Updated": "2020-08-26T07:29:38.0799707Z",
    "Created": "2020-08-25T08:23:20.4575204Z",
    "_rid": "dag+AIS5XlsyAAAAAAAAAA==",
    "_self": "dbs/dag+AA==/colls/dag+AIS5Xls=/docs/dag+AIS5XlsyAAAAAAAAAA==/",
    "_attachments": "attachments/",
    "_ts": 1598426978
}
```

**Comment by marcschier** [comment]
#774

**Comment by karok2m** [comment]
Issue of duplicate `DataSetWriterId`s in a job definition has been fixed by #689. The fix is present in `2.7.199` and `2.7.206` versions of the platform.
